cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -374,69 +255,58 @@ angular.module(PKG.name + '.commons')
-        }
-        endpoints.push(node.name);
-
-        var type = GLOBALS.pluginConvert[node.type];
-
-        switch(type) {
-          case 'source':
-            vm.instance.addEndpoint(node.name, sourceOrigin, {uuid: 'Origin' + node.name});
-            vm.instance.addEndpoint(node.name, sourceTarget, {uuid: 'Target' + node.name});
-            break;
-          case 'sink':
-            vm.instance.addEndpoint(node.name, sinkOrigin, {uuid: 'Origin' + node.name});
-            vm.instance.addEndpoint(node.name, sinkTarget, {uuid: 'Target' + node.name});
-            break;
-          case 'action':
-            vm.instance.addEndpoint(node.name, actionOrigin, {
-              uuid: 'Origin' + node.name,
-              cssClass: node.type + '-anchor'
-            });
-            vm.instance.addEndpoint(node.name, actionTarget, {
-              uuid: 'Target' + node.name,
-              cssClass: node.type + '-anchor'
-            });
-            break;
-          default:
-            // Need to id each end point so that it can be used later to make connections.
-            var originId = {uuid: 'Origin' + node.name};
-            var targetId = {uuid: 'Target' + node.name};
-            if (node.plugin.name === 'Wrangler') {
-              originId.cssClass = 'wrangler-anchor';
-              targetId.cssClass = 'wrangler-anchor';
-            }
-
-            if (node.type === 'errortransform') {
-              originId.cssClass = 'error-anchor';
-              targetId.cssClass = 'error-anchor';
+        let sourceObj = {
+          isSource: true,
+          filter: function(event) {
+            // we need this variable because when the user clicks on the endpoint circle, multiple
+            // 'mousedown' events are fired
+            if (event.target.className === 'endpoint-circle' && !endpointClicked) {
+              endpointClicked = true;
             }
 
-            vm.instance.addEndpoint(node.name, transformOrigin, originId);
-            vm.instance.addEndpoint(node.name, transformTarget, targetId);
-            break;
+            return event.target.className === 'endpoint-circle';
+          },
+          connectionType: 'basic'
+        };
+        if (vm.isDisabled) {
+          sourceObj.enabled = false;
         }
+        vm.instance.makeSource(node.name, sourceObj);
+
+        vm.instance.makeTarget(node.name, {
+          isTarget: true,
+          dropOptions: { hoverClass: 'drag-hover' },
+          anchor: 'ContinuousLeft',
+          allowLoopback: false
+        });
       });
     }
 
     function addConnections() {
       angular.forEach($scope.connections, function (conn) {
-        var sourceNode = $scope.nodes.filter( node => node.name === conn.from);
-        var targetNode = $scope.nodes.filter( node => node.name === conn.to);
-        if (!sourceNode.length || !targetNode.length) {
-          return;
+        var sourceNode, targetNode;
+
+        for (let i = 0; i < $scope.nodes.length; i++) {
```

Comment:
I think we could use array's `find` method. I think the reason we went with imperative approach is to not go through the entire `node` array twice I believe but at the max we will have 50-100 nodes and filtering through that should be nearly `O(1)` in a modern browser. Declarative `find` or `filter` is easier to read and is more concise too.